File: haskell/src/Variables.hs
module Variables
    ( demonstrateVariables
    ) where

-- | Function to demonstrate various variable types in Haskell
demonstrateVariables :: IO ()
demonstrateVariables = do
    putStrLn "=== Variables ==="
    
    -- Basic types
    let integerVar = 42 :: Integer
        doubleVar = 3.14159 :: Double
        floatVar = 2.718 :: Float
        charVar = 'A' :: Char
        stringVar = "Hello, World!" :: String
        boolVar = True :: Bool
    
    putStrLn $ "Integer: " ++ show integerVar
    putStrLn $ "Double: " ++ show doubleVar
    putStrLn $ "Float: " ++ show floatVar
    putStrLn $ "Char: " ++ show charVar
    putStrLn $ "String: " ++ stringVar
    putStrLn $ "Boolean: " ++ show boolVar
    
    -- Tuples
    let point2D = (10, 20) :: (Int, Int)
        point3D = (1, 2, 3) :: (Int, Int, Int)
        person = ("Alice", 25) :: (String, Int)
    
    putStrLn "\nTuples:"
    putStrLn $ "2D Point: " ++ show point2D
    putStrLn $ "3D Point: " ++ show point3D
    putStrLn $ "Person: " ++ show person
    
    -- Lists
    let numbers = [1..5] :: [Int]
        words' = ["apple", "banana", "orange"] :: [String]
        chars = ['a'..'e'] :: [Char]
    
    putStrLn "\nLists:"
    putStrLn $ "Numbers: " ++ show numbers
    putStrLn $ "Words: " ++ show words'
    putStrLn $ "Characters: " ++ show chars
    
    -- Maybe type
    let justValue = Just 42 :: Maybe Int
        nothingValue = Nothing :: Maybe Int
    
    putStrLn "\nMaybe types:"
    putStrLn $ "Just value: " ++ show justValue
    putStrLn $ "Nothing value: " ++ show nothingValue
    
    -- Custom data type
    data Color = Red | Green | Blue deriving Show
    let color = Blue :: Color
    
    putStrLn "\nCustom type:"
    putStrLn $ "Color: " ++ show color
    
    -- Type aliases
    type Age = Int
    type Name = String
    let age = 25 :: Age
        name = "Bob" :: Name
    
    putStrLn "\nType aliases:"
    putStrLn $ "Name: " ++ name
    putStrLn $ "Age: " ++ show age
    
    putStrLn "" -- Empty line

File: r/src/variables.R
# Function to demonstrate variables in R
demonstrate_variables <- function() {
  cat("=== Variables ===\n")
  
  # Numeric variables
  integer_var <- 42L  # Explicit integer using L suffix
  double_var <- 3.14
  
  # Character variables
  name <- "John"
  message <- 'Hello, World!'
  
  # Logical variables
  is_true <- TRUE
  is_false <- FALSE
  
  # Factor (categorical) variable
  colors <- factor(c("red", "blue", "green"))
  
  # Vector
  numbers <- c(1, 2, 3, 4, 5)
  
  # List
  person <- list(
    name = "Alice",
    age = 25,
    scores = c(90, 85, 95)
  )
  
  # Display values
  cat("Integer:", integer_var, "\n")
  cat("Double:", double_var, "\n")
  cat("Name:", name, "\n")
  cat("Message:", message, "\n")
  cat("Is True:", is_true, "\n")
  cat("Is False:", is_false, "\n")
  cat("Colors (Factor):", levels(colors), "\n")
  cat("Numbers (Vector):", numbers, "\n")
  cat("Person (List):\n")
  str(person)
  cat("\n")
}